src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -740,27 +736,52 @@ private static void RegisterGrainReferenceSerializers(Type type)
                 GrainReference.SerializeGrainReference,
                 (expected, stream) =>
                 {
-                    ConstructorInfo constructor;
-                    if (expected.IsConstructedGenericType)
+                    Func<GrainReference, GrainReference> ctorDelegate;
+                    var deserialized = (GrainReference)GrainReference.DeserializeGrainReference(expected, stream);
+                    if (expected.IsConstructedGenericType == false)
                     {
-                        var referenceType = type.MakeGenericType(expected.GenericTypeArguments);
-                        constructor =
-                            referenceType.GetConstructor(
-                                BindingFlags.CreateInstance | BindingFlags.NonPublic | BindingFlags.Instance,
-                                null,
-                                new[] { typeof(GrainReference) },
-                                null);
+                        return defaultCtorDelegate(deserialized);
                     }
-                    else
+
+                    if (grainRefConstructorDictionary.TryGetValue(expected, out ctorDelegate) == false)
```

Comment:
nit: I know the new Star Wars just came out, but Yoda notation is less readable than simply writing `!condition`.